pkg/disttask/framework/scheduler/scheduler_manager.go

```diff
@@ -405,3 +409,94 @@ func (sm *Manager) MockScheduler(task *proto.Task) *BaseScheduler {
 		slotMgr: sm.slotMgr,
 	})
 }
+
+func (sm *Manager) collectMetricsLoop() {
+	logutil.Logger(sm.ctx).Info("collect metrics loop start")
+	ticker := time.NewTicker(defaultCollectMetricsInterval)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-sm.ctx.Done():
+			logutil.BgLogger().Info("collect metrics loop exits")
+			return
+		case <-ticker.C:
+			sm.collectMetrics()
+		}
+	}
+}
+
+func (sm *Manager) collectMetrics() {
+	subtasks, err := sm.taskMgr.GetAllSubtasks(sm.ctx)
+	if err != nil {
+		logutil.BgLogger().Warn("get all subtasks failed", zap.Error(err))
+		return
+	}
+	allNodes, err := sm.taskMgr.GetAllNodes(sm.ctx)
+	if err != nil {
+		logutil.BgLogger().Warn("get all nodes failed", zap.Error(err))
+		return
+	}
+	// taskID => execID => state => cnt
+	subtaskCnt := make(map[int64]map[string]map[proto.SubtaskState]int)
+	taskType := make(map[int64]proto.TaskType)
+	for _, subtask := range subtasks {
+		if _, ok := subtaskCnt[subtask.TaskID]; !ok {
+			subtaskCnt[subtask.TaskID] = make(map[string]map[proto.SubtaskState]int, len(allNodes))
+			for _, node := range allNodes {
```

Comment:
some subtask might have a execID that's not in allNodes, such as a node is dead, but it's subtasks haven't balance away.

Reply:
prefer we init the map as need, can also reduce a `sql`

Reply:
Yes, so I prefer to recording the dead node in meta db. If we do not record it, we We have no other way to delete previous metrics.